kyaml/yaml/rnode.go

```diff
@@ -147,6 +179,164 @@ func NewMapRNode(values *map[string]string) *RNode {
+// struct, a slice or a yaml.Node.
+func NewRNodeFrom(val interface{}) (*RNode, error) {
+	rn, err := func() (*RNode, error) {
+		if val == nil {
+			return nil, fmt.Errorf("the passed-in object must not be nil")
+		}
+
+		switch typedVal := val.(type) {
+		case *RNode:
+			return typedVal, nil
+		case RNode:
+			return &typedVal, nil
+		case *Node:
+			return &RNode{value: typedVal}, nil
+		case Node:
+			return &RNode{value: &typedVal}, nil
+		case []*RNode:
+			if len(typedVal) == 0 {
+				return NewListRNode(), nil
+			}
+			return rnodesToRNode(typedVal), nil
+		}
+
+		kind := reflect.ValueOf(val).Kind()
+		if kind == reflect.Ptr {
+			kind = reflect.TypeOf(val).Elem().Kind()
+		}
+
+		switch kind {
+		case reflect.Struct, reflect.Map:
+			node, err := newMappingRNodeFromTypedObject(val)
+			if err != nil {
+				return nil, err
+			}
+			return node, nil
+		case reflect.Slice:
+			node, err := newSequenceRNodeFromTypedObject(val)
+			if err != nil {
+				return nil, err
+			}
+			return node, nil
+		case reflect.String:
+			var s string
+			switch val := val.(type) {
+			case string:
+				s = val
+			case *string:
+				s = *val
+			}
+			return NewStringRNode(s), nil
+		case reflect.Int, reflect.Int64:
+			var i int
+			switch val := val.(type) {
+			case int:
+				i = val
+			case *int:
+				i = *val
+			case int64:
+				i = int(val)
+			case *int64:
+				i = int(*val)
+			}
+			return NewIntRNode(i), nil
+		case reflect.Float64:
+			var f float64
+			switch val := val.(type) {
+			case float64:
+				f = val
+			case *float64:
+				f = *val
+			}
+			return NewFloatRNode(f), nil
+		case reflect.Bool:
+			var b bool
+			switch val := val.(type) {
+			case bool:
+				b = val
+			case *bool:
+				b = *val
+			}
+			return NewBoolRNode(b), nil
+		default:
+			return nil, fmt.Errorf("unsupported kind %s", kind)
+		}
+	}()
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert typed object to RNode: %w", err)
+	}
+	return rn, nil
+}
+
+func newMappingRNodeFromTypedObject(v interface{}) (*RNode, error) {
+	// The built-in types only have json tags. We can't simply do ynode.Encode(v),
+	// since it use the lowercased field name by default if no yaml tag is specified.
+	// This affects both k8s built-in types (e.g. appsv1.Deployment) and any types
+	// that depends on built-in types (e.g. metav1.ObjectMeta, corev1.PodTemplate).
+	// To work around it, we rely on the json tags. We first convert v to
+	// map[string]interface{} through json and then convert it to ynode.
+	node, err := func() (*yaml.Node, error) {
+		j, err := json.Marshal(v)
+		if err != nil {
+			return nil, err
+		}
+		var m map[string]interface{}
+		if err = json.Unmarshal(j, &m); err != nil {
+			return nil, err
+		}
+
+		node := &yaml.Node{}
+		if err = node.Encode(m); err != nil {
+			return nil, err
+		}
+		// Encode set the Style to FlowStyle sometimes, we reset the Style to empty.
+		node.Style = 0
+		return node, sortFields(node)
```

Comment:
I don't think we should make assumptions about / mess with the contents of the node in such a low-level function. Please remove this sorting and the `cleanupCreationTimestamp` piece.